fix OpenTracing trace context propagation with Fault Tolerance @Asynchronous methods

Ladicek opened the pull request:
There used to be a test for that, but didn't involve `@Asynchronous`
methods, so was in fact useless. This commit changes the test to
actually call an `@Asynchronous` method, which revealed an issue.
The `smallrye-fault-tolerance-tracing-propagation` dependency was
missing, so trace context propagation didn't actually work.

Then, the test also uncovered an issue in SmallRye Fault Tolerance
(see https://github.com/smallrye/smallrye-fault-tolerance/pull/208),
which is why this commit also brings a new version of it.

In addition to that, this commit also fixes a devmode issue where
on hot reload, the Jaeger extension tried to register a gauge
that was already present. That ended up with an exception.

And finally, this commit also adds a test for tracing JDBC.

Resolves #7981.
Resolves #7982.
Resolves #7984.

Ladicek committed:
fix OpenTracing trace context propagation with Fault Tolerance @Asynchronous methods

There used to be a test for that, but didn't involve `@Asynchronous`
methods, so was in fact useless. This commit changes the test to
actually call an `@Asynchronous` method, which revealed an issue.
The `smallrye-fault-tolerance-tracing-propagation` dependency was
missing, so trace context propagation didn't actually work.

Then, the test also uncovered an issue in SmallRye Fault Tolerance
(see https://github.com/smallrye/smallrye-fault-tolerance/pull/208),
which is why this commit also brings a new version of it.

In addition to that, this commit also adds a test for tracing JDBC.

Ladicek commented on extensions/smallrye-opentracing/deployment/pom.xml:
```diff
@@ -30,6 +30,20 @@
         </dependency>
+        <dependency>
+            <groupId>io.smallrye</groupId>
+            <artifactId>smallrye-fault-tolerance-tracing-propagation</artifactId>
```
So this is what I'd like to ask about. I'm not exactly sure where to add this dependency so that it also works in user applications (it obviously works in the unit test, which is in the same module :-) ).

The intent is that this module is available when both `quarkus-smallrye-fault-tolerance` and `quarkus-smallrye-opentracing` are present.

gsmet replied:
/cc @aloubyansky @stuartwdouglas 

I don't think we have anything for that. Not sure how we could solve this.

Ladicek replied:
Some extensions already bring other extensions transitively (e.g. I think Fault Tolerance brings Context Propagation). We can do the same here, but not sure from which side :-)

aloubyansky replied:
@Ladicek by work in user applications you mean as a runtime dependency?

To this day, runtime dependencies remain in user's hands, expressed in traditional maven/gradle way. And e.g. `dependency:tree` (and the gradle equivalents) is still showing the actual runtime dependency tree. From this perspective, it's not like we don't have anything for that, it's the driving build system.
We could definitely come up with an enhancement to be able to stick other dependencies into the runtime classpath, it will break what's mentioned above though.

I think today, you'd have to either add it as a dependency of the runtime artifact or ask users to add it explicitly as an application dependency.

aloubyansky replied:
> Some extensions already bring other extensions transitively (e.g. I think Fault Tolerance brings Context Propagation). We can do the same here, but not sure from which side :-)

The least popular one :)

Ladicek replied:
Hi @michalszynkiewicz @pavolloffay @loicmathieu, do you have an opinion here? Would it be better for Fault Tolerance to automatically bring OpenTracing, or the other way around?

michalszynkiewicz replied:
@Ladicek isn't a mistake (my mistake :D) that `smallrye-fault-tolerance-tracing-propagation` depends on `smallrye-fault-tolerance`?
If we could drop this dependency, we don't need to bring the whole fault tolerance with opentracing, right?

Ladicek replied:
Yea, correct. In fact, just moving the dependency on `smallrye-fault-tolerance-tracing-propagation` to the `runtime` module should do, no? It brings some unnecessary dependencies on SmallRye FT, but those should basically be ignored, because the Quarkus extension isn't present. Or is that a complete misunderstanding?

stuartwdouglas replied:
This is actually a relatively common use case, so I think we should add something to handle it in our resolution/classloading infrasturucture.

There are lots of cases like this where you only want an integration module if two different dependencies are present, I am not 100% sure what the best way to express/handle this would be though.

loicmathieu replied:
> It brings some unnecessary dependencies on SmallRye FT, but those should basically be ignored, because the Quarkus extension isn't present.

In JVM mode the JAR will be included inside the final artifact by Maven. We do nothing about it. I'm not sure we can prune some dependencies out of the build easily.

You should use one of these two Maven tricks to avoid pulling all dependencies :
- Add your dependency as optional: it's what we do for health and metrics, not sure it can be used here as you want to include this one if both are present ...
- Exclude all transitive dependencies that will be included if the other libary (fault tolerance) is included.

Then use the `Capabilities` facility to check if both FT and context-propagation are present and add the functionality inside the Quarkus build. But apparently the fonctionality is automatically added thanks to the presence of the dependency so this can be tricky.

I think the best is what @michalszynkiewicz suggest: make `smallrye-fault-tolerance-tracing-propagation` not depending on `smallrye-fault-tolerance` if possible.

Ladicek replied:
Oh, dependency exclusions should work. Basically, `quarkus-smallrye-opentracing` would depend on `smallrye-fault-tolerance-tracing-propagation` (but that dependency would exclude all transitive dependencies). It only contains 1 class and a `ServiceLoader` file for MP Context Propagation. Hmm, I wonder what would happen if this would be present, ConProp would be present, but FT would be missing... Well, since `smallrye-fault-tolerance-tracing-propagation` doesn't really depend on anything FT, we would probably do just the right thing -- have tracing context propagated. I need to test this.

(Rant: OpenTracing needs to integrate with ConProp, then we could remove this from FT entirely. We need ConProp in MP Platform!)

Ladicek replied:
OK, that works, I've updated the PR.

Ladicek replied:
I mean: I have made `smallrye-fault-tolerance-tracing-propagation` a dependency of `quarkus-smallrye-opentracing`, but all transitive dependencies are excluded. This works both when Fault Tolerance is present and missing. This is because `smallrye-fault-tolerance-tracing-propagation` in fact integrates MP OpenTracing with MP Context Propagation, and it only lives in SRye Fault Tolerance by accident.

I've tested that in an external test suite; adding tests to Quarkus would require adding 1 or 2 Maven modules, which is something I'd like to avoid.

Ladicek commented:
Hmm, the failing test passes for me... I'll dig into it.

Ladicek commented:
Hmm, it seems the test is quite flaky :-(

Ladicek commented:
I can reproduce the test failure locally when I let the test repeat 1000 times. It typically takes cca 400 attempts for the failure to occur. Not terribly fun to debug and I suspect there's some thread safety issue buried deep in ... I don't know where :-)

gsmet commented:
Not sure if it makes sense to backport that one, given the tests are flaky?

Ladicek commented:
I would _love_ to have this backported, but the test issue indeed needs to be resolved first.

Ladicek commented:
I tracked this down to a SmallRye Fault Tolerance threading issue: https://github.com/smallrye/smallrye-fault-tolerance/pull/208 Will update this PR when new SmallRye Fault Tolerance is released.

Ladicek commented:
This is now ready.

The failing test is unrelated: `io.quarkus.mongodb.reactive.legacy.AxleReactiveMongoClientTest`.

The test that was previously failing (`TracingTest`) is now OK. I've ran it many thousands of times locally while figuring out what's wrong (there's was a concurrency issue in SmallRye Fault Tolerance which is fixed in new version).

gsmet commented:
@jmartisk could you have a look? Thanks!

jmartisk commented on extensions/jaeger/runtime/src/main/java/io/quarkus/jaeger/runtime/QuarkusJaegerMetricsFactory.java:
```diff
@@ -45,12 +46,17 @@ public void durationMicros(long time) {
-        JaegerGauge gauge = registry.register(meta(name, MetricType.GAUGE), new JaegerGauge(), toTagArray(tags));
+        Tag[] tagArray = toTagArray(tags);
+        JaegerGauge gauge = (JaegerGauge) registry.getGauges().get(new MetricID(name, tagArray));
+        if (gauge == null) {
```
After https://github.com/quarkusio/quarkus/issues/7018 we automatically drop and re-create all metric registries during shutdown before a hot reload, so it should not happen anymore that the gauge is still registered. Are you sure we still need this? That would mean the registry dropping is wrong?!

Ladicek replied:
I've seen an exception during hot reload on this. I'll try again.

Ladicek replied:
No, this is still required. When I revert this change, I get this on hot reload:

```
org.jboss.resteasy.spi.UnhandledException: java.lang.RuntimeException: No reflection exceptions should be thrown unless there is a fundamental error in your code set up.
        at org.jboss.resteasy.core.ExceptionHandler.handleException(ExceptionHandler.java:381)
        at org.jboss.resteasy.core.SynchronousDispatcher.writeException(SynchronousDispatcher.java:216)
        at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:515)
        at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:259)
        at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:160)
        at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:362)
        at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:163)
        at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:245)
        at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:73)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:122)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.access$000(VertxRequestHandler.java:36)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler$1.run(VertxRequestHandler.java:87)
        at io.quarkus.runtime.CleanableExecutor$CleaningRunnable.run(CleanableExecutor.java:231)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
        at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2027)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1551)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1442)
        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
        at java.lang.Thread.run(Thread.java:748)
        at org.jboss.threads.JBossThread.run(JBossThread.java:479)
Caused by: java.lang.RuntimeException: No reflection exceptions should be thrown unless there is a fundamental error in your code set up.
        at io.jaegertracing.internal.metrics.Metrics.createMetrics(Metrics.java:76)
        at io.jaegertracing.internal.metrics.Metrics.<init>(Metrics.java:32)
        at io.jaegertracing.internal.metrics.Metrics.<init>(Metrics.java:28)
        at io.jaegertracing.Configuration.getTracerBuilder(Configuration.java:221)
        at io.quarkus.jaeger.runtime.QuarkusJaegerTracer.tracer(QuarkusJaegerTracer.java:40)
        at io.quarkus.jaeger.runtime.QuarkusJaegerTracer.buildSpan(QuarkusJaegerTracer.java:59)
        at io.opentracing.util.GlobalTracer.buildSpan(GlobalTracer.java:133)
        at io.quarkus.smallrye.opentracing.runtime.TracerProducer_ProducerMethod_tracer_96dadb3d6afa0cccadfe742c3e06ad433737c844_ClientProxy.buildSpan(TracerProducer_ProducerMethod_tracer_96dadb3d6afa0cccadfe742c3e06ad433737c844_ClientProxy.zig:238)
        at io.opentracing.contrib.jaxrs2.server.ServerTracingFilter.filter(ServerTracingFilter.java:69)
        at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:310)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:439)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:400)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:374)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:67)
        at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:488)
        ... 20 more
Caused by: java.lang.IllegalArgumentException: A metric with metricID MetricID{name='jaeger_tracer_reporter_queue_length', tags=[]} already exists
        at io.smallrye.metrics.MetricsRegistryImpl.register(MetricsRegistryImpl.java:131)
        at io.quarkus.jaeger.runtime.QuarkusJaegerMetricsFactory.createGauge(QuarkusJaegerMetricsFactory.java:48)
        at io.jaegertracing.internal.metrics.Metrics.createMetrics(Metrics.java:66)
        ... 34 more
```

Ladicek replied:
@jmartisk Any idea? Thanks! :-)

jmartisk replied:
That's fishy, I'll try to reproduce this and let you know

jmartisk replied:
@Ladicek I can't reproduce that - I created a simple app which uses tracing and metrics, called one traced method, triggered a reload, called that traced method again, the `jaeger_tracer_reporter_queue_length` gauge was re-registered correctly and no errors occurred.
I used Quarkus built from this PR branch, but with this one change reverted. Any other specific thing that I should try? :)

Ladicek replied:
Oh well. I've just verified this still occurs, even if I rebase this PR on top of current `master`. I'll put together a detailed reproducer.

Ladicek replied:
@jmartisk Here's a reproducer with detailed instructions: https://github.com/Ladicek/quarkus-jaeger-metrics-hot-reload-issue

jmartisk replied:
@Ladicek  Ha! Thanks for the reproducer. I think I understand why this happens now and why it didn't happen with my reproducer: your application does **not** have a dependency on the `quarkus-smallrye-metrics` module, so the metrics processor doesn't run and doesn't register the task to clean up registries during reload.

But what now? Is it correct that Jaeger registers metrics even if the application does not declare a dependency on `quarkus-smallrye-metrics`? In other extensions, we generally try to make their dependency on metrics optional, so metrics are activated only if the user explicitly adds the  `quarkus-smallrye-metrics` dependency.

jmartisk replied:
And if we can't make changes to the Jaeger extension such that metrics become optional (and we can run without them completely), then I suppose we should make the Jaeger module depend on `quarkus-smallrye-metrics-deployment` in order to make sure that the metrics processor is always applied and can do proper cleanup

jmartisk replied:
It also happens that if the Metrics extension is not present (as in your reproducer), then metrics are not being exposed on the `/metrics` endpoint, so they are quite useless.

Ladicek replied:
Now that is ... interesting. `quarkus-jaeger` depends on `smallrye-metrics` and `microprofile-metrics-api`, but not on `quarkus-smallrye-metrics`. That is ... wrong, for sure.

Ladicek replied:
Or perhaps that was done to make metrics semi-optional...

Ladicek replied:
@jmartisk What's the way to detect if metrics are enabled? I guess `capabilities.isCapabilityPresent(Capabilities.METRICS)`? If so, then we could decide on that and add either `NoopMetricsFactory` or `QuarkusJaegerMetricsFactory`. And probably make the dependencies on `smallrye-metrics` and `microprofile-metrics-api` `<scope>provided</scope>`, because they are only present to be able to compile `QuarkusJaegerMetricsFactory`. Does that sound about right? If so, I can check if that works and submit a PR.

jmartisk replied:
@Ladicek yes, checking for the capability is ok.
I think the correct way is have an `optional` dependency on `quarkus-smallrye-metrics`, like Agroal does it:
https://github.com/quarkusio/quarkus/blob/master/extensions/agroal/runtime/pom.xml#L44

jmartisk replied:
And what would be absolutely optimal and consistent with most other extensions, would be that the metrics from the extension are disabled by default (**even if** the metrics dependency is present) and have to be enabled by a config property `quarkus.jaeger.metrics.enabled`. See https://quarkus.io/guides/writing-extensions#extension-metrics

Even more optimal would be using the Metrics SPI (mentioned in the linked document) to register metrics eagerly, but if you feel that's too much for now, I guess we can leave that for later :)

Ladicek replied:
Seriously? :open_mouth: :grin:

Ladicek replied:
I mean, I would probably be able to do all that, but that would make this PR pretty large. I just wanted to fix the trace context propagation, found a dev mode issue, included a fix which was tiny, and now this :-/

Here's my proposal: I already have finished the conditional inclusion/exclusion of metrics based on the capability. I have also changed `quarkus-jaeger` to no longer depend on `smallrye-metrics` and `microprofile-metrics-api`; instead, it uses an `optional` dependency on `quarkus-smallrye-metrics` as you suggest. It seems to work (didn't try native :grin:). What's is _not_ done is addressing your 2nd comment, the one starting with "And what would be absolutely optimal" -- I'd leave that for future. Agree?

jmartisk replied:
Sure, that's clearly not the goal of this PR. 
I created https://github.com/quarkusio/quarkus/issues/8502 for that

Ladicek replied:
Nice, thanks! I'll update this PR in a few minutes.

Ladicek replied:
PR rebased and updated per discussion above.

Ladicek committed:
fix Jaeger metrics collection when SmallRye Metrics is absent

The Jaeger extension used to collect metrics unconditionally,
even when the SmallRye Metrics extension wasn't present.
With this commit, the Jaeger extension uses `NoopMetricsFactory`
if SmallRye Metrics are missing, and only uses the real
metrics factory if SmallRye Metrics are present.

This is not totally consistent with how metrics are enabled/disabled
in all other extensions. Fixing that is left for future.

jmartisk reviewed (approved):
LGTM!

gsmet reviewed (approved):
Approving, per Jan.
